scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,245 @@
+  private Map<String, String> actToId;
+
+  //boolean stating whether the simulation results should be regenerated
+  private boolean areLastSimResultsDirty;
+
+  //simulation results so far
+  private SimulationResults lastSimResults;
+  //cached simulation results cover the period [Duration.ZERO, lastSimResultsEnd]
+  private Duration lastSimResultsEnd = Duration.ZERO;
+
+  //map from activity start time to serialized activity and its name
+  private List<SimulatedActivity> activitiesInserted = new ArrayList<>();
+
+  record SimulatedActivity(Duration start, SerializedActivity activity, String name) implements Comparable<SimulatedActivity> {
+    @Override
+    public int compareTo(@NotNull final SimulatedActivity o) {
+      return start.compareTo(o.start);
+    }
+  }
+
+  public IncrementalSimulationDriver(MissionModel<?> missionModel){
+    this.missionModel = missionModel;
+    initSimulation();
+  }
+
+  private void initSimulation(){
+    actToId = new HashMap<>();
+    lastSimResults = null;
+    this.engine = new SimulationEngine();
+    activitiesInserted.clear();
+
+    /* The top-level simulation timeline. */
+    this.timeline = new TemporalEventSource();
+    areLastSimResultsDirty = false;
+    this.cells = new LiveCells(timeline, missionModel.getInitialCells());
+    /* The current real time. */
+    curTime = Duration.ZERO;
+
+    // Begin tracking all resources.
+    for (final var entry : missionModel.getResources().entrySet()) {
+      final var name = entry.getKey();
+      final var resource = entry.getValue();
+      engine.trackResource(name, resource, curTime);
+    }
+
+    // Start daemon task(s) immediately, before anything else happens.
+    {
+      final var daemon = engine.initiateTaskFromSource(missionModel::getDaemon);
+      final var commit = engine.performJobs(Set.of(SimulationEngine.JobId.forTask(daemon)),
+                                            cells, curTime, Duration.MAX_VALUE, missionModel);
+      timeline.add(commit);
+    }
+  }
+
+  public void simulateActivity(SerializedActivity activity, Duration startTime, String nameAct){
+    var testAct = new SimulatedActivity(startTime, activity, nameAct);
+    if(startTime.noLongerThan(curTime)){
+      var toBeInserted = new ArrayList<>(activitiesInserted);
+      toBeInserted.add(testAct);
+      initSimulation();
+      simulateManyActs(toBeInserted);
+    } else {
+      simulateAct(testAct);
+    }
+  }
+
+
+  /**
+   * Get the simulation results from the Duration.ZERO to the current simulation time point
+   * @return the simulation results
+   */
+  public SimulationResults getSimulationResults(){
+    return getSimulationResultsUntil(curTime);
+  }
+
+  /**
+   * Get the simulation results from the Duration.ZERO to a specified end time point.
+   * The provided simulation results might cover more than the required time period.
+   * @return the simulation results
+   */  public SimulationResults getSimulationResultsUntil(Duration endTime){
```

Comment:
This method's first line has rolled up onto the end of its javadoc.
```suggestion
   */
   public SimulationResults getSimulationResultsUntil(Duration endTime){
```